### PR TITLE
Don't interpret effects returned from fork

### PR DESCRIPTION
**Pull request by restrry**
| Q                        | A  |
| ------------------------ | ---  |
| Fixed Issues?            | https://github.com/redux-saga/redux-saga/issues/1130|
| Patch: Bug Fix?          |  bug fix   |
| Major: Breaking Change?  |  yes  |
| Minor: New Feature?      |    |
| Tests Added + Pass?      | yes |
| Any Dependency Changes?  |    |

PR to start discussion with code examples.
straightforward way to prevent effect from execution. ¯\_(ツ)_/¯

### CODE REVIEW

File: packages/core/src/internal/proc.js [hunk 1]
```diff
@@ -22,7 +22,7 @@ import { getLocation, addSagaStack, sagaStackToString } from './error-utils'
 import { asap, suspend, flush } from './scheduler'
 import { channel, isEnd } from './channel'
 import matcher from './matcher'
-
+const DONT_INTERPRET_EFFECT = Symbol('DONT_INTERPRET_EFFECT')
```
Comment (restrry) [inline packages/core/src/internal/proc.js]: could be moved to symbol library if we decide that the current solution is ok

**Comment by Andarist** [comment]
Thanks for starting a discussion around this but the solution itself seems to be hacky (brute forcing the expected result) in its current state.

The whole problem lies in how `createTaskIterator` wraps the forked simple functions (not generators). If we try to comment out this part of the (refactored) code - https://github.com/redux-saga/redux-saga/blob/7895e4be95e414cf814068c9a2c490370dd07eb2/packages/core/src/internal/proc.js#L124-L135 we'll get 3 failing tests:
- [saga fork/join handling : functions](https://github.com/redux-saga/redux-saga/blob/cb31736fcd9bcf701d6388acb4687128e07ab649/packages/core/__tests__/interpreter/forkjoin.js#L96])
- [middleware's custom emitter](https://github.com/redux-saga/redux-saga/blob/cb31736fcd9bcf701d6388acb4687128e07ab649/packages/core/__tests__/middleware.js#L79)
- [saga detached forks failures](https://github.com/redux-saga/redux-saga/blob/cb31736fcd9bcf701d6388acb4687128e07ab649/packages/core/__tests__/interpreter/forkJoinErrors.js#L144)

By replacing the quoted code with this:
```js
return makeIterator(() => ({ done: true, value: result }))
```
the only still failing test is the first mentioned one.

Tracing it to its roots we can find those commits:
- [test itself](https://github.com/redux-saga/redux-saga/commit/3c93f01b0ea1a866e01ca3e06fde15722fa3df7c#diff-aef8708b040ddfd27e50ad40d8b3e146R458)
- [tested implementation](https://github.com/redux-saga/redux-saga/commit/7adac320c23f7e20568fd30a1c6b9bfe8a81b49f#diff-80f482071554cef3f7a8b76903c8de16R89)

The question is what is supposed to be a "simple effect" here? What actual use cases are covered by this behaviour of forked function returning an interpretable effect?

**Comment by restrry** [comment]
seems that the case when fork could return any kind of effect that we can interpret - including promise and iterator,
https://github.com/redux-saga/redux-saga/blob/433c333053ba5b1a3187406b09078e9e4ee99e73/packages/core/src/internal/proc.js#L334-L336
at least such a test for promise was added here https://github.com/redux-saga/redux-saga/commit/3c93f01b0ea1a866e01ca3e06fde15722fa3df7c#diff-aef8708b040ddfd27e50ad40d8b3e146R473

**Comment by Andarist** [comment]
> seems that the intention was to cover a case when fork returns any kind of effect that we can interpret, especially a promise and an iterator.

Yeah, I suppose the promises are the most useful use case. Dropping support for this feature would force this:
```js
const fn = () => delayP(200).then(() => 'foo')
// ...
const task = yield fork(fn)
// ...
const result = yield join(task)
```
to be refactored to
```js
const _fn = () => delayP(200).then(() => 'foo')
function* fn() { return yield call(_fn) }

// ...
const task = yield fork(fn)
// ...
const result = yield join(task)
```

Seems like the current behaviour for *promises* (and maybe iterators, although questionable they both probably should be handled the same way because they are in other parts of the code) is useful.

No other effect should be "flattened" like this, because it is not flattened in other scenarios:
- when using generator syntax and returning an effect descriptor
- when using a call effect on a function returning an effect descriptor

If we agree on that behaviour the mentioned part of the code could be rewritten like this:
```js
 let shouldResolve = is.promise(result) || is.iterator(result)
  
 const next = (value = result) => { 
   if (shouldResolve) { 
     shouldResolve = false 
     return { done: false, value: result } 
   } else { 
     return { done: true, value } 
   } 
 } 
  
 return makeIterator(next) 
```

**Comment by Andarist** [comment]
Would u like to rebase this, implement my proposal & check if this is covered nicely by tests?

**Comment by restrry** [comment]
@Andarist sure thing

File: packages/core/src/internal/proc.js [hunk 2]
```diff
@@ -121,14 +121,11 @@ function createTaskIterator({ context, fn, args }) {
...
     }
 
-    let resolved = false
-
-    const next = arg => {
-      if (!resolved) {
-        resolved = true
-        return { done: false, value: result }
+    const next = (value = result) => {
+      if (is.promise(value)) {
```
[the remark this answers is not in the document]
Comment (restrry) [reply packages/core/src/internal/proc.js]: case with iterator is covered here https://github.com/redux-saga/redux-saga/pull/1573/files#diff-ecefd26543e2e37544ae6d726d23b588R120 
no needs to wrap in another iterator

File: packages/core/__tests__/interpreter/fork.js [hunk 3]
```diff
@@ -0,0 +1,78 @@
+import { createStore, applyMiddleware } from 'redux'
+import sagaMiddleware from '../../src'
+import * as io from '../../src/effects'
+
+describe('should not interpret returned effect. issue #1130', () => {
```
[the remark this answers is not in the document]
Comment (restrry) [reply packages/core/__tests__/interpreter/fork.js]: only in babel plugin, ok I can change to `test` in sake of consistency

**Comment by shinima** [comment]
LGTM, except the order of value and done as @Andarist says:

> also probably better to reorder value and done, value should comes first, because its the order used by v8

**Comment by shinima** [comment]
I think we can add some warning if users are trying to return effects in forked functions like `yield io.fork(() => io.select())`. But it is of low priority though.

File: packages/core/src/internal/proc.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (Andarist) [reply packages/core/src/internal/proc.js]: cool, havent noticed it

**Pull request merged**